crates/engine/primitives/src/lib.rs

```diff
@@ -42,6 +42,27 @@ pub use invalid_block_hook::InvalidBlockHook;
 pub mod config;
 pub use config::*;
 
+/// Trait for accessing tree state during validation
+pub trait TreeStateProvider: Send + Sync {
+    /// Returns the current canonical block hash
+    fn canonical_block_hash(&self) -> alloy_primitives::B256;
+
+    /// Returns the current canonical block number  
+    fn canonical_block_number(&self) -> u64;
+}
+
+/// Context providing access to tree state during validation
+pub struct TreeCtx<'a> {
+    /// The engine API tree state
+    pub state: &'a dyn TreeStateProvider,
+}
```

Comment:
we likely need a generic for N as well because I think we need to fetch parent block for example

Reply:
makes sense, added a generic `State` which later we can specify in impls as `EngineApiTreeState<N: NodePrimitives>`